### PR TITLE
mgr/dashboard: Add 'Browse Dashboards' button in multi-cluster and ceph-cluster Grafana dashboards

### PR DESCRIPTION
**Pull request by piyushagarwal1411**
Fixes: https://tracker.ceph.com/issues/68316

Signed-off-by:- Piyush Agarwal <piyushagarwal14.pa@gmail.com> 

[Screencast from 2024-10-03 14-39-22.webm](https://github.com/user-attachments/assets/c8c41def-7c58-4acb-84df-f98766ecb943)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by aaSharma14** [comment]
jenkins test make check

File: monitoring/ceph-mixin/dashboards/utils.libsonnet [hunk 1]
```diff
@@ -133,6 +133,30 @@ local timeSeries = import 'timeseries_panel.libsonnet';
...
+    icon: icon,
+    includeVars: includeVars,
+    keepTime: keepTime,
+    tags: tags,
+    targetBlank: targetBlank,
+    title: title,
+    tooltip: tooltip,
+    type: type,
+    url: url,
+  },
```
Comment (aaSharma14) [inline monitoring/ceph-mixin/dashboards/utils.libsonnet:157]:
```suggestion
  addLinkSchema(asDropdown,
                icon,
                includeVars,
                keepTime,
                tags,
                targetBlank,
                title,
                tooltip,
                type,
                url)::
    {
      asDropdown: asDropdown,
      icon: icon,
      includeVars: includeVars,
      keepTime: keepTime,
      tags: tags,
      targetBlank: targetBlank,
      title: title,
      tooltip: tooltip,
      type: type,
      url: url,
    },
```

File: monitoring/ceph-mixin/dashboards/multi-cluster.libsonnet [hunk 2]
```diff
@@ -43,6 +43,22 @@ local g = import 'grafonnet/grafana.libsonnet';
...
+    includeVars=true,
+    keepTime=true,
+    tags=[],
+    targetBlank=false,
+    title='Browse Dashboards',
+    tooltip='',
+    type='dashboards',
+    url=''
+  )
+])
```
Comment (aaSharma14) [inline monitoring/ceph-mixin/dashboards/multi-cluster.libsonnet:59]:
```suggestion
    .addLinks([
      $.addLinkSchema(
        asDropdown=true,
        icon='external link',
        includeVars=true,
        keepTime=true,
        tags=[],
        targetBlank=false,
        title='Browse Dashboards',
        tooltip='',
        type='dashboards',
        url=''
      )
    ])
```

**Review by aaSharma14 — CHANGES_REQUESTED** [review]

File: monitoring/ceph-mixin/dashboards/multi-cluster.libsonnet [hunk 3]
```diff
@@ -43,6 +43,22 @@ local g = import 'grafonnet/grafana.libsonnet';
...
+        tags=[],
+        targetBlank=false,
+        title='Browse Dashboards',
+        tooltip='',
+        type='dashboards',
+        url=''
+      )
+    ])
+
+
```
Comment (aaSharma14) [inline monitoring/ceph-mixin/dashboards/multi-cluster.libsonnet:61]:
```suggestion
```

Avoid adding extra blank lines

File: monitoring/ceph-mixin/dashboards/utils.libsonnet [hunk 4]
```diff
@@ -133,6 +133,30 @@ local timeSeries = import 'timeseries_panel.libsonnet';
...
+      keepTime: keepTime,
+      tags: tags,
+      targetBlank: targetBlank,
+      title: title,
+      tooltip: tooltip,
+      type: type,
+      url: url,
+    },
+
+
```
Comment (aaSharma14) [inline monitoring/ceph-mixin/dashboards/utils.libsonnet:159]:
```suggestion
```

Same here

**Comment by aaSharma14** [comment]
jenkins retest this please

**Comment by piyushagarwal1411** [comment]
jenkins test make check

**Comment by piyushagarwal1411** [comment]
jenkins retest this please

**Comment by aaSharma14** [comment]
jenkins test api

**Review by aaSharma14 — APPROVED** [review]
LGTM

**Comment by piyushagarwal1411** [comment]
jenkins test api

File: src/pybind/mgr/dashboard/ci/check_grafana_dashboards.py [hunk 5]
```diff
@@ -103,8 +103,6 @@ def get_grafana_dashboards(base_dir):
                 title = dashboard_config['title']
                 assert len(title) > 0, \
                     "Title not found in '{}'".format(json_file)
-                assert len(dashboard_config.get('links', [])) == 0, \
-                    "Links found in '{}'".format(json_file)
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/ci/check_grafana_dashboards.py:107]: What's the reason behind this change? Why do we need to remove this assert?

**Review by Pegonzal — COMMENTED** [review]
Thanks for the PR @piyushagarwal1411!
Couple questions from my side:
- Why is the button only for multi-cluster and ceph-cluster dashboards? As seen in the video demo when accessing any other dashboard the button disappers, doesn't seem right

File: src/pybind/mgr/dashboard/ci/check_grafana_dashboards.py (hunk 5 above)
Comment (piyushagarwal1411) [reply src/pybind/mgr/dashboard/ci/check_grafana_dashboards.py:107]: @Pegonzal It is checking whether the links field in the Grafana dashboard configuration is empty or not,  and if it does, the script will raise an error, flagging that the json_file contains links. We tried creating an exception for the files in which the button was added, but it was still giving error in make check, so me and @aaSharma14 just decided to drop that assert as it is not needed.

Also i just added the button for multi-cluster and ceph-cluster dashboards in this PR since it was my first PR and i was just figuring out how it works. I will be adding it to the rest of the dashboards and will create a separate PR for the same.

**Comment by aaSharma14** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test api

File: src/pybind/mgr/dashboard/ci/check_grafana_dashboards.py (hunk 5 above)
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/ci/check_grafana_dashboards.py:107]: > just decided to drop that assert as it is not needed.

If this was not really needed I was trying to verify why was this here in the first place,
and traced it back to [this PR](https://github.com/ceph/ceph/pull/31951) by @epuertat, but could not find an explanation for the changes in that PR. Which, actually is just removing exactly what we are doing currently in this PR.
So unless @epuertat has any reason for it, this is okay with me as I don't have any.

> Also i just added the button for multi-cluster and ceph-cluster dashboards in this PR since it was my first PR and i was just figuring out how it works. I will be adding it to the rest of the dashboards and will create a separate PR for the same.

Sure, that's fine with me, ideally you could create a tracker so we don't miss on doing this, and if you already know you are going to work on it, just assign it to yourself and move it forward.

**Comment by aaSharma14** [comment]
jenkins test api

File: src/pybind/mgr/dashboard/ci/check_grafana_dashboards.py (hunk 5 above)
Comment (piyushagarwal1411) [reply src/pybind/mgr/dashboard/ci/check_grafana_dashboards.py:107]: Yeah i checked that PR by @epuertat as well. Do we really need that assert or can we go ahead with removing it? @epuertat 

And I'll create the tracker @Pegonzal

**Review by cloudbehl — APPROVED** [review]
LGTM
